src/main/java/org/candlepin/model/ConsumerCurator.java

```diff
@@ -1001,6 +1006,180 @@ else if (systemUuid != null) {
+        if (consumerIds == null || consumerIds.isEmpty()) {
+            return new ArrayList<>();
+        }
+
+        String retrieveIdCertsHql = "SELECT idCert.id FROM Consumer WHERE id IN (:consumerIds)";
+        return entityManager.get()
+            .createQuery(retrieveIdCertsHql)
+            .setParameter("consumerIds", consumerIds)
+            .getResultList();
+    }
+
+    /**
+     * Retrieves the content access certificate ids for provided consumer ids.
+     *
+     * @param consumerIds - ids of the {@link Consumer}s.
+     * @return content access certificate ids.
+     */
+    public List<String> getContentAccessCertIds(Collection<String> consumerIds) {
+        if (consumerIds == null || consumerIds.isEmpty()) {
+            return new ArrayList<>();
+        }
+
+        String retrieveScaCertsHql = "SELECT contentAccessCert.id FROM Consumer WHERE id IN (:consumerIds)";
+        return entityManager.get()
+            .createQuery(retrieveScaCertsHql)
+            .setParameter("consumerIds", consumerIds)
+            .getResultList();
+    }
+
+    /**
+     * Retrieves all the serial ids for provided {@link ContentAccessCertificate} ids
+     * and {@link IdentityCertificate} ids.
+     *
+     * @param caCertIds - ids of content access certificates.
+     * @param idCertIds - ids of identity certificates.
+     * @return serial ids.
+     */
+    public List<Long> getSerialIdsForCerts(Collection<String> caCertIds, Collection<String> idCertIds) {
+        List<Long> serialIds = new ArrayList<>();
+        if (caCertIds != null && !caCertIds.isEmpty()) {
+            String caCertHql =
+                "SELECT ca.serial.id " +
+                "FROM ContentAccessCertificate ca " +
+                "WHERE ca.id IN (:certIds)";
+
+            serialIds.addAll(entityManager.get()
+                .createQuery(caCertHql)
+                .setParameter("certIds", caCertIds)
+                .getResultList());
+        }
+
+        if (idCertIds != null && !idCertIds.isEmpty()) {
+            String idCertHql =
+                "SELECT idcert.serial.id " +
+                "FROM IdentityCertificate idcert " +
+                "WHERE idcert.id IN (:certIds)";
+
+            serialIds.addAll(entityManager.get()
+                .createQuery(idCertHql)
+                .setParameter("certIds", idCertIds)
+                .getResultList());
+        }
+
+        return serialIds;
+    }
+
+    /**
+     * Deletes {@link Consumer}s based on the provided consumer ids.
+     *
+     * @param consumerIds - ids of the consumers to delete.
+     * @return the number of consumer that were deleted.
+     */
+    public int deleteConsumers(Collection<String> consumerIds) {
+        if (consumerIds == null || consumerIds.isEmpty()) {
+            return 0;
+        }
+
+        Collection<Consumer> consumersToDelete = getConsumers(consumerIds);
+        super.bulkDelete(consumersToDelete);
+        return consumersToDelete.size();
```

Comment:
With the spec test I realized that deleting all of the consumers with a single HQL statement based on the consumer ids causes FK constraint violations for the following FKs for Consumer.

This isn't ideal as we have to load all the consumers in memory and individually delete each Consumer with org.hibernate.Session.delete() so that the cascading cleans up all of the FK constraints. I wasn't able to find another way to handle this with Hibernate, but I could be missing something. The alternative is to have methods for each constraint that deletes them before deleting the actual consumer rows. 

**FK Constraints for Consumer**
1. facts
2. environmentIds
3. activationKeys
4. hypervisorId
5. capabilities
6. guestIds
7. installedProducts
8. keyPairData
9. Entitlements

Reply:
Ugh, I should have seen this during design :/
The third, ideal alternative would be to add `ON DELETE CASCADE` on the foreign keys in the db themselves, which we have a card for in the backlog. This feature is where it would be ideal to have that done, but I am conflicted on doing this now because it would slow this task down.

Reply:
We should investigate if CASCADE even is a viable solution. From what I've seen it suffers from the same performance issues as deleting with hibernate. If it really is just as slow we would have to implement bulk delete of all these separate items ourselves of live with hibernate solution.

Reply:
Adding the `ON DELETE CASCADE` to the foreign keys makes a lot of sense. I added the change in a new change set 20220613000000-add_delete_cascade_to_consumer_fks.

Reply:
I'd also like to have more data on the performance implications, but my general stance on them are to use them within reason.

I also agree that object linkage and DB triggers or triggered behavior needs to be limited to logical objects rather than just done on every link. In this case, that means if we delete a consumer, it also makes sense to delete the things that make up the consumer's children (facts, caps, tags, etc.), but NOT where other objects point to it. In those cases we need to make a decision on a case-by-case basis depending on how critical it is to both track when such a thing goes missing (i.e. pool -> entitlement). Setting a field to null rather than a full cascade might also make more sense in some instances.

Reply:
Since I see at least 2 people have concerns about performance, and also about how exactly we should go about it (cascade delete vs setting to null), then I think we should leave the cascade as a separate investigation, and go with bulk deletes of all these separate tables on this task.

Reply:
It makes sense that we should only adding cascades on the fields that are owned by the consumers and I updated the change set to only include the FKs from the cp_consumer_content_tags, cp_consumer_facts, cp_consumer_capability, and  cp_consumer_guests tables.

I also did some performance testing by creating 100k inactive consumers and 100k active consumers, and then performing the InactiveConsumerCleanerJob with the FKs having the cascade deletes. The average time to delete a batch of 1000 inactive consumers was 72.71649 ms and the total time to complete the job was 13266 ms.

I was also curious to see what the performance for using the Hibernate deletion without having the cascade on delete on the FKs and it is pretty bad and was taking over 10,000 ms to delete a batch of 1000 inactive consumers on the same data set. I didn't finish the job as it was taking a long time!

Reply:
While not fully a representative perf test (there would have to be ~millions of rows in all related tables to say it is), given that I see we already have cascades for some children tables of consumer (cp_consumer_environments, cp_consumer_activation_key, cp_consumer_hypervisor), I'd say let's go for it.

The only thing is that we need to add 2 more cascades for the cp_consumer_guests_attributes (child table of  cp_consumer_guests) and cp_installed_products tables (child table of cp_consumer).

Reply:
Added the cascades for the FK on cp_consumer_guests_attributes table and the cp_installed_products table.